### PR TITLE
Improve docs around knn similarity search

### LINKED ISSUE #102924: Update docs for search_knn_with_expected_similarity
Knn with expected similarity needs to be clarified and updated:

 - We need to indicate that "similarity" here is the similarity BEFORE score transformations
 - Indicate an appropriate transformation for a given score for each of our vector similarity functions
 - Needs to be backported to 8.8 (note, any Max-inner-product examples only go back to 8.11).

https://www.elastic.co/guide/en/elasticsearch/reference/current/knn-search.html#_search_knn_with_expected_similarity

### PR DESCRIPTION
**Pull request by benwtrent**
Adding equations to the docs around how to best calculate similarity & score. The similarity parameter for search was added in 8.8.

The max-inner-product mentions will be removed for all versions before 8.11 when backporting.

closes: https://github.com/elastic/elasticsearch/issues/102924

### CODE REVIEW

File: docs/reference/search/search-your-data/knn-search.asciidoc [hunk 1]
```diff
@@ -563,6 +564,18 @@ minimum similarity for a vector to be considered a match. The `knn` search flow
 * Do not return any vectors that are further away than the configured `similarity`
 --
 
+NOTE: `similarity` is the true <<dense-vector-similarity, similarity>> before it has been transformed into `_score` and boost applied.
```
Comment (jdconrad) [inline docs/reference/search/search-your-data/knn-search.asciidoc:567]: I find this statement a bit confusing. Are we saying the below functions are not the "true" similarity here?

**Review by jdconrad — APPROVED** [review]
LGTM. Added a single comment that I think the new note is a bit confusing and may need some additional explanation.

File: docs/reference/search/search-your-data/knn-search.asciidoc (hunk 1 above)
Comment (benwtrent) [reply docs/reference/search/search-your-data/knn-search.asciidoc:567]: I am saying the below functions are how to transform a `_score` back into similarity if you want to change it. 

We take vector similarity and munge it into a score in all sorts of ways.

File: docs/reference/search/search-your-data/knn-search.asciidoc [hunk 2]
```diff
@@ -563,6 +564,18 @@ minimum similarity for a vector to be considered a match. The `knn` search flow
...
+
+For each configured <<dense-vector-similarity, similarity>>, here is the corresponding inverted `_score` function.
```
Comment (benwtrent) [inline docs/reference/search/search-your-data/knn-search.asciidoc:569]:
```suggestion
For each configured <<dense-vector-similarity, similarity>>, here is the corresponding inverted `_score` function. This is so if you are wanting to filter from a `_score` perspective, you can do this minor transformation to correctly reject irrelevant results.
```

**Pull request merged**